icu4c/source/tools/upropdump/upropdump.cpp

```diff
@@ -0,0 +1,178 @@
+// © 2016 and later: Unicode, Inc. and others.
+// License & terms of use: http://www.unicode.org/copyright.html
+
+#include <iostream>
+#include "toolutil.h"
+#include "uoptions.h"
+#include "cmemory.h"
+#include "charstr.h"
+#include "unicode/uchar.h"
+#include "unicode/errorcode.h"
+#include "unicode/uniset.h"
+#include "unicode/putil.h"
+#include "unicode/umutablecptrie.h"
+#include "writesrc.h"
+
+U_NAMESPACE_USE
+
+/*
+ * Global - verbosity
+ */
+UBool VERBOSE = FALSE;
+UBool QUIET = FALSE;
+
+UBool haveCopyright = TRUE;
+UCPTrieType trieType = UCPTRIE_TYPE_SMALL;
+
+void handleError(const ErrorCode& status, const char* context) {
+    if (status.isFailure()) {
+        std::cerr << "Error: " << context << ": " << status.errorName() << std::endl;
+        exit(status.get());
+    }
+}
+
+void dumpBinaryProperty(UProperty uproperty, FILE* f) {
+    ErrorCode status;
+    const char* fullPropName = u_getPropertyName(uproperty, U_LONG_PROPERTY_NAME);
+    const char* shortPropName = u_getPropertyName(uproperty, U_SHORT_PROPERTY_NAME);
+    const USet* uset = u_getBinaryPropertySet(uproperty, status);
+    handleError(status, fullPropName);
+
+    fputs("[unicode_set.data]\n", f);
+    fprintf(f, "long_name = \"%s\"\n", fullPropName);
+    usrc_writeUnicodeSet(f, shortPropName, uset, UPRV_TARGET_SYNTAX_TOML);
+}
+
+void dumpEnumeratedProperty(UProperty uproperty, FILE* f) {
+    ErrorCode status;
+    const char* fullPropName = u_getPropertyName(uproperty, U_LONG_PROPERTY_NAME);
+    const char* shortPropName = u_getPropertyName(uproperty, U_SHORT_PROPERTY_NAME);
+    const UCPMap* umap = u_getIntPropertyMap(uproperty, status);
+    handleError(status, fullPropName);
+
+    fputs("[code_point_map.data]\n", f);
+    fprintf(f, "long_name = \"%s\"\n", fullPropName);
+    usrc_writeUCPMap(f, shortPropName, umap, uproperty, UPRV_TARGET_SYNTAX_TOML);
+    fputs("\n", f);
+
+    LocalUMutableCPTriePointer builder(umutablecptrie_fromUCPMap(umap, status));
+    LocalUCPTriePointer utrie(umutablecptrie_buildImmutable(
+        builder.getAlias(),
+        trieType,
+        UCPTRIE_VALUE_BITS_32, // TODO(review): What is the best way to pick the value width?
```

Comment:
Please call [u_getIntPropertyMaxValue(uproperty)](https://unicode-org.github.io/icu-docs/apidoc/released/icu4c/uchar_8h.html#a4a379ff77d7f9e0717ceeb7703b08df3) and see if it's up to 0xff or 0xffff.

Please also `U_ASSERT(u_getIntPropertyMinValue(uproperty) >= 0);`

Reply:
13e28462bbed57b95ef20be3af1b5ed474d4b174